sync/validation_queue.go

```
@@ -44,149 +67,146 @@ func (vq *validationQueue) inQueue(id types.BlockID) bool {
 
-func (vq *validationQueue) addDependencies(blk *types.BlockHeader, checkDatabase func(id types.BlockID) (*types.Block, error)) bool {
+func (vq *blockQueue) addDependencies(jobId interface{}, blks []types.BlockID, finishCallback func(res bool) error) (bool, error) {
+	vq.Lock()
+	vq.callbacks[jobId] = finishCallback
 	dependencys := make(map[types.BlockID]struct{})
-	for _, id := range blk.ViewEdges {
+	idsToPush := make([]types.Hash32, 0, len(blks))
+	for _, id := range blks {
 		if vq.inQueue(id) {
-			vq.reverseDepMap[id] = append(vq.reverseDepMap[id], blk.ID())
-			vq.Info("add block %v to %v dependencies map", id, blk.ID())
+			vq.reverseDepMap[id] = append(vq.reverseDepMap[id], jobId)
+			vq.Info("add block %v to %v pending map", id, jobId)
 			dependencys[id] = struct{}{}
 		} else {
 			//	check database
-			if _, err := checkDatabase(id); err != nil {
+			if _, err := vq.GetBlock(id); err != nil {
 				//unknown block add to queue
-				vq.queue <- id
-				vq.reverseDepMap[id] = append(vq.reverseDepMap[id], blk.ID())
-				vq.Info("add block %v to %v dependencies map", id, blk.ID())
+				vq.reverseDepMap[id] = append(vq.reverseDepMap[id], jobId)
+				vq.Info("add block %v to %v pending map", id, jobId)
 				dependencys[id] = struct{}{}
+				idsToPush = append(idsToPush, id.AsHash32())
 			}
 		}
 	}
+	vq.Unlock()
 
-	vq.depMap[blk.ID()] = dependencys
-	return len(dependencys) > 0
-}
-
-func (vq *validationQueue) addToDatabase(id types.BlockID) error {
-	if callback, ok := vq.callbacks[id]; ok {
-		if err := callback(); err != nil {
-			return err
-		}
+	for _, id := range idsToPush {
+		vq.addToPending([]types.Hash32{id})
 	}
-	return nil
-}
 
-func (vq *validationQueue) getMissingBlocks() []types.BlockID {
-	missingBlocks := make([]types.BlockID, 0, len(vq.reverseDepMap))
-	for k := range vq.reverseDepMap {
-		missingBlocks = append(missingBlocks, k)
+	//todo better this is a little hacky
+	if len(dependencys) == 0 {
+		return false, finishCallback(true)
```

IMO it is better that you'll handle the error that might be returned by `finishCallback(true)` here. This way `addDependencies` won't need to return any error and it will simplify the caller of this method (which currently need to "remember" that this error actually means that the block had no dependencies and that the execution of its callback failed.

---

its called from the sync for a layer and the validation queue each needs to know the error because they handle it differently